app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -10,15 +13,49 @@ def wordpress_hosted
     end
   end
 
-  def setup
+  def saml
+    @user = User.find_for_saml(request.env["omniauth.auth"], current_user)
+    if @user.persisted?
+      flash[:notice] = I18n.t("devise.omniauth_callbacks.success", :kind => $DOMAIN_CONFIG['saml_provider'])
+      sign_in_and_redirect @user, :event => :authentication #this will throw if @user is not activated
+    else
+      session["devise.saml_auth_data"] = request.env["omniauth.auth"]
+      redirect_to new_user_registration_url
+    end
+  end
+
+  def setup_wordpress_hosted
+    @tenant = Tenant.current_tenant
+    return unless allow_wordpress_oauth_authentication?(@tenant)
+    
+    if @tenant.present?
+      Rails.logger.info("**********Setup Started for #{@tenant.database_name}************")
+      Rails.logger.info("The tenant auth key is #{@tenant.wp_auth_key}")
+      Rails.logger.info("The tenant secret is #{@tenant.wp_auth_secret}")
+      Rails.logger.info("The tenant url is #{@tenant.wp_url}")
+    end
+
+    request.env['omniauth.strategy'].options[:client_id] = wp_auth_key(@tenant)
+    request.env['omniauth.strategy'].options[:client_secret] = wp_auth_secret(@tenant)
+    request.env['omniauth.strategy'].options[:client_options].site = wp_url(@tenant)
+    render :text => "Setup complete.", :status => 404
+  end
+
+  def setup_saml
     @tenant = Tenant.current_tenant
-    Rails.logger.info("**********Setup Started for #{@tenant.database_name}************")
-    Rails.logger.info("The tenant auth key is #{@tenant.wp_auth_key}")
-    Rails.logger.info("The tenant secret is #{@tenant.wp_auth_secret}")
-    Rails.logger.info("The tenant url is #{@tenant.wp_url}")
-    request.env['omniauth.strategy'].options[:client_id] = @tenant.wp_auth_key.present? ? @tenant.wp_auth_key : ENV["WP_AUTH_KEY"]
-    request.env['omniauth.strategy'].options[:client_secret] = @tenant.wp_auth_secret.present? ? @tenant.wp_auth_secret : ENV["WP_AUTH_SECRET"]
-    request.env['omniauth.strategy'].options[:client_options].site = @tenant.wp_url.present? ? @tenant.wp_url : ENV["WP_URL"]
+    return unless allow_saml_authentication?(@tenant)
+
+    if @tenant.present?
+      Rails.logger.info("**********Setup Started for #{@tenant.database_name}************")
+      Rails.logger.info("The tenant auth key is #{@tenant.wp_auth_key}")
+      Rails.logger.info("The tenant secret is #{@tenant.wp_auth_secret}")
+      Rails.logger.info("The tenant url is #{@tenant.wp_url}")
+    end
+
+    request.env['omniauth.strategy'].options[:idp_cert_fingerprint] = idp_cert_fingerprint(@tenant)
+    request.env['omniauth.strategy'].options[:idp_sso_target_url] = idp_sso_target_url(@tenant)
+    request.env['omniauth.strategy'].options[:issuer] = url_for(action: 'metadata', controller: 'saml')
```

Comment:
@jamiefolsom - Added this to include the `issuer` property as `<current-tentant-url>/sp/metadata`.

Example: `https://purdue-staging.covecollective.org/sp/metadata`.

Reply:
Makes sense, although I am now getting a 404 response for [that request](https://purdue-staging.covecollective.org/sp/metadata): 
<img width="1150" alt="Screen Shot 2020-09-23 at 09 55 52" src="https://user-images.githubusercontent.com/882377/94022481-21edd800-fd83-11ea-9c96-03016725da9a.png">

Reply:
Hmm, interesting. I haven't pushed anything to staging yet. I'll take a look.